lib/api/api.js

```diff
@@ -184,14 +186,101 @@ const api = {
             }
             return { returnTagCount, isImplicitDeny };
         }
+        let bb;
+        let fileEventData = null;
+
+        if (apiMethod === 'objectPost') {
+            bb = busboy({ headers: request.headers });
+        }
 
         return async.waterfall([
+            next => {
+                if (apiMethod === 'objectPut' || apiMethod === 'objectPutPart') {
+                    return next(null);
+                }
+                if (apiMethod === 'objectPost' && request.headers['content-type'].includes('multipart/form-data')) {
+                    writeContinue(request, response);
+
+                    request.formData = {};
+                    let totalFieldSize = 0;
+                    const MAX_FIELD_SIZE = 20 * 1024; // 20KB
+                    bb.on('field', (fieldname, val) => {
+                        totalFieldSize += Buffer.byteLength(val, 'utf8');
+                        if (totalFieldSize > MAX_FIELD_SIZE) {
+                            const err = errors.MaxPostPreDataLengthExceeded;
+                            bb.emit('error', err);
+                        }
+                        // Convert fieldname to lowercase for case-insensitive comparison
+                        const lowerFieldname = fieldname.toLowerCase();
+                        request.formData[lowerFieldname] = val;
+                    });
+
+                    bb.on('file', (fieldname, file, filename, encoding, mimetype) => {
+                        fileEventData = { fieldname, file, filename, encoding, mimetype };
+                        return next(null);
+                    });
+
+                    bb.on('finish', () => {
+                        // if fields are found but no file, continue
+                        if (!fileEventData) {
+                            return next(null);
+                        }
+                        return undefined;
+                    });
+
+                    bb.on('error', (err) => {
+                        log.trace('Error processing form data:', {
+                            error: err,
+                        });
+                        request.unpipe(bb);
+                        return next(err);
+                    });
+
+                    request.pipe(bb);
+                } else {
+                    // issue 100 Continue to the client
+                    writeContinue(request, response);
+                    const MAX_POST_LENGTH = request.method === 'POST' ?
+                        1024 * 1024 : 1024 * 1024 / 2; // 1 MB or 512 KB
+                    const post = [];
+                    let postLength = 0;
+                    request.on('data', chunk => {
+                        postLength += chunk.length;
+                        // Sanity check on post length
+                        if (postLength <= MAX_POST_LENGTH) {
+                            post.push(chunk);
```

Comment:
In this case we are now reading data and building the `post` buffer before checking authorization. Is that ok ?

Reply:
This is the handler for POST requests other than those of `objectPost` , contained within the `else` condition; I have maintained it as-is per the existent behaviour in the product and thought better than to tamper with it, as that may change behaviour for a handful of APIs.
To be fair to it, it also has a limit on the length of 1Mb, so I don't see this as an issue.

Reply:
This is no longer relevant as the flows have been separated